llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h

```diff
@@ -180,31 +186,45 @@ class VPBuilder {
 
   VPValue *createNot(VPValue *Operand, DebugLoc DL = {},
                      const Twine &Name = "") {
+    if (auto *V = Folder.foldNot(Operand))
+      if (BB)
+        return getOrAddLiveIn(V);
     return createInstruction(VPInstruction::Not, {Operand}, DL, Name);
   }
 
   VPValue *createAnd(VPValue *LHS, VPValue *RHS, DebugLoc DL = {},
                      const Twine &Name = "") {
+    if (auto *V = Folder.foldAnd(LHS, RHS))
```

Comment:
I am not sure if it is possible, but it would be nice if the folder could just have a single interface that takes opcode + operands and then we can have a single place like `createInstruction` that queries the folder, instead of having to update each create* function here and add a corresponding `foldXXX` function to the folder, which seems quite verbose.

Reply:
I thought about this, but I fear we'll have to match the VPInstruction using a switch, and dispatch appropriately: not sure if this is preferable, so let's keep this thread open.